Contributors

sophyphreak opened the pull request:
Hey, I added a simple list of all the current contributors. With this set up, you add the linkedin links manually, I know this feature isn't finished. You said you were thinking of a way to connect this up to the database?

monarchwadia commented:
Hi @sophyphreak, this is great! 

What I meant earlier was the need to put this on the backend as some kind of service, so that instead of calling Github directly, it calls an API that we have built on our own.

This makes the code a lot more consistent, maintainable, and changeable. It decouples the frontend from a direct dependency on Github, and allows us to implement caching for this data on the backend.

Can you please create the following endpoint in the backend, and hit it from the frontend?

```
GET /api/v1/contributors
```

For now, the endpoint can return Mintbean-specific contributors. In the future, we can add an `:id` parameter to make it return issues for specific projects

monarchwadia commented:
@sophyphreak I'm not sure how familiar you are with the backend -- please let me know if I can help you understand more :-)

sophyphreak commented:
Can do. I'll let you know if I have any trouble. Also, I'm surprised about the merge conflict. I should be able to fix that. Thanks!

sophyphreak commented:
Github seems to believe there are no merge conflicts, so I think you fixed that problem when you merged my branch with the master branch. 

I added a route to the backend to get the contributors info from the Github api. Everything seems to work, but the styling in the contributors component is different due to a change in the CSS. It functions, however, so I decided to leave it how it is. 

Is there anything more I can work on with this component, or should I move onto something else in the project? I'm happy to help however I can.

sophyphreak commented:
@monarchwadia, any problems?

monarchwadia commented:
Hi @sophyphreak , checking this now!

monarchwadia commented on backend/route/contributors.route.js:
```diff
@@ -0,0 +1,206 @@
+    //             .then(attendance => attendance.update(request.payload));
+    //     }
+    // });
+};
```
Could you please remove the commented out code? Commented-out code should never be committed. [Google search](https://www.google.com/search?q=commented+out+code+should+not+be+committed&oq=commented+out+code+should+not+be+committed&aqs=chrome..69i57.4862j1j7&sourceid=chrome&ie=UTF-8).

sophyphreak replied:
Sorry about that. Silly mistake.

monarchwadia commented on backend/route/contributors.route.js:
```diff
@@ -0,0 +1,206 @@
+        'sophyphreak': 'https://www.linkedin.com/in/sophyphreak/'
+    }
+    return linkedIn[username] || null
+}
```
Nice. This will do for now. Once auth is in place, we'll want to move to a database-driven solution.
Can you please move the `const linkedIn` to a separate file called `contributors.js` so that we can ask contributors to upload their profile content directly inside that single file?

sophyphreak replied:
Can do.

monarchwadia commented on backend/route/contributors.route.js:
```diff
@@ -0,0 +1,206 @@
+    const linkedIn = {
+        'sophyphreak': 'https://www.linkedin.com/in/sophyphreak/'
+    }
+    return linkedIn[username] || null
```
You don't need to explicitly return `null` here. Just `return linkedIn[username]` will return `undefined` if there is no entry for the username.

sophyphreak replied:
Okay.

monarchwadia commented on frontend/src/views/Contributors.vue:
```diff
@@ -0,0 +1,56 @@
+<template lang="pug">
+  div
+    h1 Contributors
+    h2(v-if="this.errorMessage") {{ errorMessage }}
```
For SEO purposes, `h` tags should never be used for anything other than semantic titles and headers.
Can you please use a simple `div` or `p` tag with a classname here, instead?

sophyphreak replied:
Sure.

monarchwadia commented on frontend/src/views/Contributors.vue:
```diff
@@ -0,0 +1,56 @@
+<template lang="pug">
+  div
+    h1 Contributors
+    h2(v-if="this.errorMessage") {{ errorMessage }}
```
This should eventually be replaced with a global error handler component

sophyphreak replied:
Okay, noted.

monarchwadia commented on frontend/src/views/Contributors.vue:
```diff
@@ -0,0 +1,56 @@
+          td 
+            a(:href="contributor.portfolio") {{ contributor.portfolio && 'portfolio' }}
+          td
+            a(:href="contributor.linkedin") {{ contributor.linkedin && 'linkedin' }}
```
Is this table responsive on mobile?

sophyphreak replied:
It is not. I will fix it.

sophyphreak replied:
It is now.

monarchwadia commented on frontend/src/views/Contributors.vue:
```diff
@@ -0,0 +1,56 @@
+          th LinkedIn
+        tr
+          td {{ contributor.name }}
+          td 
+            a(:href="contributor.html_url") {{contributor.login}}
+          td 
+            a(:href="contributor.portfolio") {{ contributor.portfolio && 'portfolio' }}
+          td
+            a(:href="contributor.linkedin") {{ contributor.linkedin && 'linkedin' }}
```
Use v-if instead. Also, have you seen `frontend/src/components/mb-a.vue` ? It adds a bit of code that makes external links more secure and standardized.

Untested code. 

```suggestion
          td 
            mb-a(:href="contributor.html_url" :v-if="contributor.login") {{contributor.login}}
          td 
            mb-a(:href="contributor.portfolio" :v-if="contributor.portfolio") Portfolio
          td
            mb-a(:href="contributor.linkedin" :v-if="contributor.portfolio") LinkedIn
```

sophyphreak replied:
Okay.

monarchwadia commented on frontend/src/views/Contributors.vue:
```diff
@@ -0,0 +1,56 @@
+      errorMessage: null,
+    };
+  },
+  created() {
+    (async () => {
+      const apiService = new ApiService();
+      try {
+        const { data } = await apiService.get('/api/v1/contributors');
+        this.contributors = data;
+        this.errorMessage = null;
+      } catch (e) {
+        this.contributors = [];
+        this.errorMessage = "Failed to fetch contributors.";
+        console.log(e);
+      }
+    })();
+  },
```
```suggestion
  created() {
    new ApiService()
    .get('/api/v1/contributors')
    .then(({ data }) => {
      this.contributors = data;
      this.errorMessage = null;
    })
    .catch(e => {
      console.error(e);
      this.contributors = [];
      this.errorMessage = "Failed to fetch contributors";
    })
  
  },
```

sophyphreak replied:
I can do this change, but I'm not sure I understand the pros for using ES6 promises compared to using ES7 promises.

monarchwadia commented on frontend/src/views/Contributors.vue:
```diff
@@ -0,0 +1,56 @@
+.result {
+  padding: 5px;
+  margin-bottom: 16px;
+}
```
See frontend/src/styles/_dimensions.scss
use `vstep` to maintain vertical rhythm and `hstep` to maintain horizontal rhythm. You will probably have to `@import` these
Try the following values:
```scss
@import '../styles/dimensions';

.result {
  padding: vstep(1) hstep(1);
  margin-bottom: hstep(2);
}
```

sophyphreak replied:
Ok. Thanks.

carloshdelreal replied:
I really liked this pug template system.

sophyphreak commented:
In the process of making requested changes.

sophyphreak commented:
I'm getting an error from the github server. 
```
{

    "message": "API rate limit exceeded for 73.155.178.206. (But here's the good news: Authenticated requests get a higher rate limit. Check out the documentation for more details.)",
    "documentation_url": "https://developer.github.com/v3/#rate-limiting"

}
```
The docs say we have a limit of 5000 requests per hour. I assume someone working on this went over that?

monarchwadia commented:
> I'm getting an error from the github server.
> 
> ```
> {
> 
>     "message": "API rate limit exceeded for 73.155.178.206. (But here's the good news: Authenticated requests get a higher rate limit. Check out the documentation for more details.)",
>     "documentation_url": "https://developer.github.com/v3/#rate-limiting"
> 
> }
> ```
> 
> The docs say we have a limit of 5000 requests per hour. I assume someone working on this went over that?

@sophyphreak  I think that refers to your particular IP address (which could include multiple users)

sophyphreak commented:
Okay. Makes sense. Thanks!

sophyphreak commented:
Hey @monarchwadia, I believe I've finished all the requested changes, but let me know if this needs more work before it can be merged into master. Thanks for the thorough code review and specific directions. It made my part a lot easier.

carloshdelreal commented on backend/route/contributors.route.js:
```diff
@@ -0,0 +1,32 @@
+                for (const contributor of contributors) {
+                    const { data } = await axios.get(contributor.url);
+                    contributor.name = data.name;
+                    contributor.portfolio = data.blog;
```
In the portfolio link we have to ensure that the link begins with `http://` because otherwise the link it is going to be taken as an url within the mintbean.io domain, that's why the porfolio links are pointing to a page that does not exist within the mintbean.io domain and not to the actual portfolio.

We can do this on the frontend as well but I definitely prefer to do it on this line.

I suggest to create a function
```suggestion
                    contributor.portfolio = prefixHttp(data.blog);
```
And the definition of this function check if the url is an empty string so do nothing, as well as if begins with http://, if not prefix http://. that should be enough to solve this broken url

sophyphreak replied:
I made this change, but I'm not sure it's required. And I'm betting it may add more headaches than it solves. @monarchwadia, do you have an opinion here?

sophyphreak replied:
I added a function that adds `http://` if it isn't already there.

carloshdelreal commented on backend/shared/contributors.js:
```diff
@@ -0,0 +1,5 @@
+const linkedIn = {
+    'sophyphreak': 'https://www.linkedin.com/in/sophyphreak/'
+}
```
add the current contributors
```suggestion
    'monarchwadia': 'https://www.linkedin.com/in/monarchwadia/',
    'leanlunio': 'https://www.linkedin.com/in/leanjunio/',
    'shohei-shibata:' 'https://www.linkedin.com/in/shohei-shibata-53867a41/',
    'carloshdelreal': 'https://www.linkedin.com/in/carloshdelreal/',

```

carloshdelreal commented on backend/shared/contributors.js:
```diff
@@ -0,0 +1,5 @@
+const linkedIn = {
+    'sophyphreak': 'https://www.linkedin.com/in/sophyphreak/'
```
add the current contributors
```suggestion
    'sophyphreak': 'https://www.linkedin.com/in/sophyphreak/'
    'monarchwadia': 'https://www.linkedin.com/in/monarchwadia/',
    'leanlunio': 'https://www.linkedin.com/in/leanjunio/',
    'shohei-shibata:' 'https://www.linkedin.com/in/shohei-shibata-53867a41/',
    'carloshdelreal': 'https://www.linkedin.com/in/carloshdelreal/',

```

carloshdelreal commented:
The problem that I saw with the table of contributors is in small devices

sophyphreak commented:
@carloshdelreal, thank you for your feedback. Unfortunately, I'm having some trouble understanding some of your suggestions, and I am requesting clarification. 

You said,

> I have to say that there is a lot to do in the frontend for readability, could you try to do it better??

I'm sorry to say this is a very nonspecific request. Can you be more specific?

> the blue table border should enclose each of the contributors data to highlight them and to give a sense of separate entities. As it is, just encloses all the contributors on a single blue box and it's very difficult to identify different elements.

The blue border is inherited from global CSS values, so I assumed it's continuing some theme. Can you be more specific about how you believe I can improve readability?

sophyphreak commented:
@carloshdelreal 

> The problem that I saw with the table of contributors is in small devices

I was following the advice from this article for building mobile responsive tables. Do you have another suggestion for building tables that will work? 

https://css-tricks.com/responsive-data-tables/

sophyphreak commented:
And @monarchwadia, if you can restate and/or clarify any of @carloshdelreal's suggestions, I would definitely appreciate it. Thanks!

carloshdelreal commented:
> @carloshdelreal
> 
> > The problem that I saw with the table of contributors is in small devices
> 
> I was following the advice from this article for building mobile responsive tables. Do you have another suggestion for building tables that will work?
> 
> https://css-tricks.com/responsive-data-tables/

Personally I don't like to use tables, I definitely preffer to use Bootstrap to arrange the items in a single row, I think is easier to have control on the sizes of the elements as well as the row sizes

sophyphreak commented:
@carloshdelreal, makes sense, but I believe @monarchwadia specifically requested a table.

carloshdelreal commented:
I promise I will try to be more specific in the reviews, check this out

- Remember to run `yarn lint` to fix linting issues
- I am having an unknow element error

![Untitled](https://user-images.githubusercontent.com/8964341/81451377-fce92780-9149-11ea-84c0-cbc28f50031f.png)

carloshdelreal commented:
> @carloshdelreal, makes sense, but I believe @monarchwadia specifically requested a table.

so I will blame @monarchwadia

sophyphreak commented:
> I promise I will try to be more specific in the reviews, check this out
> 
>     * Remember to run `yarn lint` to fix linting issues
> 
>     * I am having an unknow element error
> 
> 
> ![Untitled](https://user-images.githubusercontent.com/8964341/81451377-fce92780-9149-11ea-84c0-cbc28f50031f.png)

Thank you for being more specific in reviews. 

I ran `yarn lint` on the files I changed, but it looks like a lot of other files need to be fixed as well. To keep this pull request limited to the specific feature, I'm only updating the files I was already updating.

The console error was because I put a `partial` inside the template. I've updated the code so that it doesn't use a partial, so the console error no longer appears.

I'd say this is ready to be merged into the master branch if you @carloshdelreal don't have any objections. What do you think?

sophyphreak commented on backend/route/contributors.route.js:
```diff
@@ -0,0 +1,41 @@
+    });
+};
+
+const addHttpPrefix = url => {
+    if (
+        url.toLowerCase().startsWith('http://') || url.toLowerCase().startsWith('https://')
+    ) {
+        return 'http://' + url
+    }
+    return url
+}
```
Just saw this. I'm taking a look now...

sophyphreak replied:
Does this need to return `null` if `url` is empty? Since `url` is falsy when it's set to either an empty string or `null`, I'm pretty sure the behavior on the front end will be the same.

sophyphreak replied:
Good point. Updating it now.

sophyphreak replied:
@carloshdelreal, I did it a little differently, but it should work just the same.

sophyphreak commented:
Okay, @carloshdelreal fixed the `addHttpPrefix` function.

carloshdelreal commented:
@monarchwadia Though, I would like to know how to cache the github contributing api data so that our backend doesn't have to request the data to the github api every time a user sees the contributors page and therefore potentially been banned because too many requests.

sophyphreak commented:
> @monarchwadia Though, I would like to know how to cache the github contributing api data so that our backend doesn't have to request the data to the github api every time a user sees the contributors page and therefore potentially been banned because too many requests.

Agreed

sophyphreak commented:
@monarchwadia, I believe this pull request is ready to be merged. 😊

sophyphreak commented:
@monarchwadia, any problems?

sophyphreak commented:
@monarchwadia, I'm sure you're really busy, but I just wanted to make sure this pull request doesn't go fallow.

monarchwadia commented:
Hi @sophyphreak @carloshdelreal thanks for the followups! I was going to announce in today's townhall meeting that PRs will all be merged in and deployed once authentication is done (it will be done by Sunday)

Sorry for being incommunicado, I've been very heads-down with several things, but I assure you this is on top of mind for me.
